src/Functions/array/arrayPrAUC.cpp

```diff
@@ -0,0 +1,241 @@
+                 *   d_Area = Precision_n * (TP_n - TP_{n-1})
+                 * and later divide it by (TP + FN).
+                 *
+                 * This can be done because (TP + FN) is constant and equal to total positive labels.
+                 */
+                curr_precision = static_cast<Float64>(curr_tp) / curr_p;
+                area += curr_precision * (curr_tp - prev_tp);
+                prev_tp = curr_tp;
+                prev_score = sorted_labels[i].score;
+            }
+
+            if (sorted_labels[i].label)
+                curr_tp += 1;
+            curr_p += 1;
+        }
+
+        curr_precision = curr_p > 0 ? static_cast<Float64>(curr_tp) / curr_p : 1.0;
+        area += curr_precision * (curr_tp - prev_tp);
+
+        /// If there were no positive labels, Recall did not change and the area is 0
+        if (curr_tp == 0)
+            return 0.0;
+        /// Finally, we divide by (TP + FN) to obtain the Recall
+        /// At this point we've traversed the whole curve and curr_tp = total positive labels (TP + FN)
+        return area / curr_tp;
+    }
+
+    static void vector(
+        const IColumn & scores,
+        const IColumn & labels,
+        const ColumnArray::Offsets & offsets,
+        PaddedPODArray<Float64> & result,
+        size_t input_rows_count)
+    {
+        result.resize(input_rows_count);
+
+        ColumnArray::Offset current_offset = 0;
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            auto next_offset = offsets[i];
+            result[i] = apply(scores, labels, current_offset, next_offset);
+            current_offset = next_offset;
+        }
+    }
+
+public:
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo &) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        size_t number_of_arguments = arguments.size();
+
+        if (number_of_arguments < 2 || number_of_arguments > 2)
+            throw Exception(
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 2.",
+                getName(),
+                number_of_arguments);
+
+        for (size_t i = 0; i < 2; ++i)
+        {
+            const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(arguments[i].type.get());
+            if (!array_type)
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The two first arguments for function {} must be of type Array.", getName());
+
+            const auto & nested_type = array_type->getNestedType();
+            if (!isNativeNumber(nested_type) && !isEnum(nested_type))
```

Comment:
I didn't really understand the logic behind it: shouldn't we receive only numbers as values inside the arrays? Also, this check verifies whether the nested type is not native number AND not Enum, which means it would fail if the nested type is a native number or an enum, because of the && sign